src/Features/Core/Portable/UnifiedSuggestions/UnifiedSuggestedActionsSource.cs

```diff
@@ -0,0 +1,711 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+using CodeFixGroupKey = System.Tuple<Microsoft.CodeAnalysis.Diagnostics.DiagnosticData, Microsoft.CodeAnalysis.CodeActions.CodeActionPriority, Microsoft.CodeAnalysis.CodeActions.CodeActionPriority?>;
+
+namespace Microsoft.CodeAnalysis.UnifiedSuggestions
+{
+    /// <summary>
+    /// Provides mutual code action logic for both local and LSP scenarios
+    /// via intermediate interface <see cref="IUnifiedSuggestedAction"/>.
+    /// </summary>
+    internal class UnifiedSuggestedActionsSource
+    {
+        /// <summary>
+        /// Gets, filters, and orders code fixes. Must be called from the UI thread
+        /// due to the eventual call to <see cref="CodeAction.IsApplicable(Workspace)"/>.
+        /// </summary>
+        public static ImmutableArray<UnifiedSuggestedActionSet> GetFilterAndOrderCodeFixes_MustBeCalledFromUIThread(
```

Comment:
_MustBeCalledFromUIThread operations should assert that they are actually on the right thread.  you can do this by sublcassing ForegroundThreadAffinitizedObject and using this.ASsertIsForeground.  It will catch when we screw this up.

Reply:
It doesn't look like ForegroundThreadAffinitizedObject is available in this layer. Is there an alternative we can use instead?

Reply:
cc @jasonmalinowski I *thought* that the features layer shouldn't have dependencies on the UI thread, so I don't know if there is an alternative.